File: Components/reinit_discrete.m
function [output1,output2] = reinit_discrete(varargin)
% Reinitialization for discrete problems.

mode = varargin{end};
switch mode
    case 'execute'
        Solution = varargin{1};
        Problem  = varargin{2};
        Aux      = varargin{4};

        [N,D] = size(Solution.decs);
        output1 = zeros(N,D);
        for j = 1:D
            output1(:,j) = randi([Problem.bound(1,j),Problem.bound(2,j)],N,1);
        end
        output2 = Aux;

    case 'parameter'
        % n/a

    case 'behavior'
        output1 = {'';'GS'}; % always perform global search
end

if ~exist('output1','var')
    output1 = [];
end
if ~exist('output2','var')
    output2 = [];
end
end

File: src/autooptlib/components/reinit_discrete.py
"""Reinitialization for discrete problems."""
from __future__ import annotations

from typing import Any

import numpy as np

from ._utils import ensure_rng, flex_get


def _extract_shape(solution: Any) -> tuple[int, int]:
    decs = flex_get(solution, "decs")
    if callable(decs):
        arr = np.asarray(decs())
    elif decs is not None:
        arr = np.asarray(decs)
    else:
        arr = np.asarray(solution)
    if arr.ndim != 2:
        raise ValueError("Solution must be 2-D for reinit_discrete")
    return arr.shape


def reinit_discrete(*args):
    mode = args[-1]
    if mode == "execute":
        solution = args[0]
        problem = args[1]
        aux = args[3] if len(args) > 3 else None
        rng = ensure_rng(aux)

        n, d = _extract_shape(solution)
        bound = np.asarray(flex_get(problem, "bound"), dtype=int)
        lower = bound[0]
        upper = bound[1]
        new = np.empty((n, d), dtype=int)
        for j in range(d):
            new[:, j] = rng.integers(lower[j], upper[j] + 1, size=n)
        return new, aux

    if mode == "parameter":
        return None, None

    if mode == "behavior":
        return ["", "GS"], None

    raise ValueError(f"Unsupported mode: {mode}")